src/scanpy/preprocessing/_pca/__init__.py

```diff
@@ -1,33 +1,52 @@
 from __future__ import annotations
 
 import warnings
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Literal, Union, get_args, overload
 from warnings import warn
 
 import anndata as ad
 import numpy as np
 from anndata import AnnData
 from packaging.version import Version
 from scipy.sparse import issparse
-from scipy.sparse.linalg import LinearOperator, svds
-from sklearn.utils import check_array, check_random_state
-from sklearn.utils.extmath import svd_flip
-
-from .. import logging as logg
-from .._compat import DaskArray, pkg_version
-from .._settings import settings
-from .._utils import _doc_params, _empty, is_backed_type
-from ..get import _check_mask, _get_obs_rep
-from ._docs import doc_mask_var_hvg
-from ._utils import _get_mean_var
+from sklearn.utils import check_random_state
+
+from ... import logging as logg
+from ..._compat import DaskArray, pkg_version
+from ..._settings import settings
+from ..._utils import _doc_params, _empty, is_backed_type
+from ...get import _check_mask, _get_obs_rep
+from .._docs import doc_mask_var_hvg
+from ._compat import _pca_with_sparse
 
 if TYPE_CHECKING:
+    from collections.abc import Container, Mapping
+    from typing import LiteralString, TypeVar
+
     from numpy.typing import DTypeLike, NDArray
+    from scipy import sparse
     from scipy.sparse import spmatrix
-    from sklearn.decomposition import PCA
 
     from .._utils import AnyRandom, Empty
 
+    CSMatrix = sparse.csr_matrix | sparse.csc_matrix
+
+    T = TypeVar("T", bound=LiteralString)
+
+
+SvdSolvPCADaskML = Literal["auto", "full", "tsqr", "randomized"]
+SvdSolvTruncatedSVDDaskML = Literal["tsqr", "randomized"]
+SvdSolvDaskML = Union[SvdSolvPCADaskML, SvdSolvTruncatedSVDDaskML]
+
+SvdSolvPCASklearn = Literal["auto", "full", "arpack", "randomized"]
+SvdSolvTruncatedSVDSklearn = Literal["arpack", "randomized"]
+SvdSolvPCASparseSklearn = Literal["arpack"]
```

Comment:
also iffy, `_pca_with_sparse` was able to handle `lobpcg`, upstream only supports `arpack`

Reply:
fixed